apps/inheritance_app/inheritance_auth_wallet.c

```diff
@@ -0,0 +1,215 @@
+      auth->challenge_size > CHALLENGE_SIZE_MAX) {
+    inheritance_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                           ERROR_DATA_FLOW_INVALID_QUERY);
+    delay_scr_init(ui_text_inheritance_wallet_auth_fail, DELAY_TIME);
+    return false;
+  }
+
+  return true;
+}
+
+static bool auth_wallet_get_entropy() {
+  secure_data_t msgs[1] = {0};
+  msgs[0].plain_data_size = WALLET_ID_SIZE;
+  memcpy(msgs[0].plain_data, auth->wallet_id, WALLET_ID_SIZE);
+
+  card_error_type_e status = card_fetch_encrypt_data(auth->wallet_id, msgs, 1);
+
+  delay_scr_init(ui_text_inheritance_wallet_authenticating, DELAY_SHORT);
+
+  if (status != CARD_OPERATION_SUCCESS ||
+      msgs[0].encrypted_data_size > ENTROPY_SIZE_LIMIT) {
+    return false;
+  }
+  set_app_flow_status(INHERITANCE_AUTH_WALLET_STATUS_CARD_TAPPED);
+
+  memcpy((void *)auth->entropy,
+         msgs[0].encrypted_data,
+         msgs[0].encrypted_data_size);
+  auth->entropy_size = msgs[0].encrypted_data_size;
+
+  return true;
+}
+
+static bool auth_wallet_get_pairs() {
+  mnemonic_to_seed((char *)auth->entropy, "", auth->private_key, NULL);
+  ed25519_publickey(auth->private_key, auth->public_key);
+
+  return true;
+}
+
+static bool auth_wallet_get_signature() {
+  const size_t unsigned_txn_size = auth->challenge_size + WALLET_ID_SIZE;
+  uint8_t unsigned_txn[unsigned_txn_size];
+
+  memcpy(unsigned_txn, auth->challenge, auth->challenge_size);
+  memcpy(unsigned_txn + auth->challenge_size, auth->wallet_id, WALLET_ID_SIZE);
+
+  ed25519_sign(unsigned_txn,
+               unsigned_txn_size,
+               auth->private_key,
+               auth->public_key,
+               auth->signature);
+
+  int valid = ed25519_sign_open(
+      unsigned_txn, unsigned_txn_size, auth->public_key, auth->signature);
+
+  if (0 != valid) {
+    inheritance_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                           ERROR_DATA_FLOW_INVALID_DATA);
+    delay_scr_init(ui_text_inheritance_wallet_auth_fail, DELAY_TIME);
+    return false;
+  }
+
+  return true;
+}
+
+static bool send_result() {
+  inheritance_result_t result = INHERITANCE_RESULT_INIT_ZERO;
+  result.which_response = INHERITANCE_RESULT_AUTH_WALLET_TAG;
+  result.auth_wallet.which_response =
+      INHERITANCE_AUTH_WALLET_RESPONSE_RESULT_TAG;
+  memcpy(result.auth_wallet.result.signature,
+         auth->signature,
+         sizeof(ed25519_signature));
+
+  if (auth->is_setup) {
+    memcpy(result.auth_wallet.result.public_key,
+           auth->public_key,
+           sizeof(ed25519_public_key));
+  }
+
+  inheritance_send_result(&result);
+  return true;
+}
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+void inheritance_auth_wallet(inheritance_query_t *query) {
+  auth = (auth_wallet_config_t *)cy_malloc(sizeof(auth_wallet_config_t));
```

Comment:
We can simply use malloc here since, the scope of this exists only within this function, and there are no early returns. Also add a NULL check after malloc.